docs/zephyr_02_functional_requirements.sdoc

```diff
@@ -1317,6 +1317,248 @@ As a Zephyr OS user I want to... ???,
 Architectural & interface REQ
 <<<
 
+[SECTION]
+TITLE: Counting Semaphores
+
+[REQUIREMENT]
+UID: ZEP-127
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Build-time registration for pre-kernel init phase notification
+STATEMENT: >>>
+Zephyr Counting Semaphores shall build-time register with Zephyr System Initialization for notification of running in the Zephyr Pre-Kernel initialization phase.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-128
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Count Semaphore pre-Kernel notification initializes statically defined semaphore objects
+STATEMENT: >>>
+When the Counting Semaphores service is notified of the Zephyr Pre-Kernel System Initialization phase,
+
+Zephyr Counting Semaphores shall initialize all statically defined semaphores.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-129
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore dynamic initialization
+STATEMENT: >>>
+Zephyr Counting Semaphores shall expose an interface to dynamically initialize a Counting Semaphore which requires the following information:
+
+- initial count of the Counting Semaphore; and
+- limit of the Counting Semaphore.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-130
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphores Static Object Init Failures Cause Build Errors
+STATEMENT: >>>
+While a Counting Semaphore is statically allocated,
+
+If the Counting Semaphore is configured with an invalid set of initialization parameter values,
+
+the Zephyr Build System shall fail the build and report the point of failure.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-131
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Dynamic Object Init Failures Reports Failure Reason
+STATEMENT: >>>
+While a Counting Semaphore is dynamically allocated,
+
+If the Counting Semaphore initialization is configured with an invalid set of initialization parameter values,
+
+the Counting Semaphore initialization shall report the reason for the failure.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-132
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Exposes Current Count
+STATEMENT: >>>
+When the Counting Semaphore current count is requested,
+
+Zephyr shall report the current count of the Counting Semaphore.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-133
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Current Count Request Invalid Parameter is Catastrophic Error
```

Comment:
another one, that is actually not true, no parameter validation happens in k_sem_count_get right now, this happens only in userspace and parameter validation is covered in 1 single requirement (ZEP-82) already, so no need to repeat this for each interface

Reply:
UIDs  ZEP-130, ZEP-131, ZEP-133, ZEP-135, ZEP-138, ZEP-130, ZEP-142 should all be covered by higher level global requirements to validate parameters when initializing an object.  
ZEP-82 covers parameter checking of a function call.
Don't know if there is one for parameter checking of a statically initialized object.

Reply:
I saw that it was not true. I added this to kick off discussion about capturing requirements on behavior for ALL public interfaces. The behavior is undefined when an invalid pointer (e.g., NULL) is passed in. It would make much more sense if the parameters were checked when RUNTIME_ERROR_CHECKS was enabled and disabled by NO_RUNTIME_CHECKS.

This is a case where I, as a product manufacturer, believe the code is wrong because it doesn't meet my needs.

Reply:
@Timotheous I expect you are correct that the UIDs you listed are provided at a _higher level_ by other requirements. If we don't propagate those requirements to lower level elements we will end up with MANY links to the higher level requirements. It will probably be too many to find gaps in.

IIRC, when an architecture definition process completes:
- all the requirements on the (Element) are delegated out to each of the sub-(Element)s;
- all the interactions and requirements on the interfaces are captured and traced back to the requirements on the - (Element) to verify all functionality has been met.

Though the explicit delegation of requirements to each specific element seems "repetitious" the subject of the statements changes and those responsible for the sub-(Element) do not have to dig through a mountain of higher requirements trying to decide what does and does not apply.